src/algebraic_geometry/elliptic_curve/point.lean

```diff
@@ -0,0 +1,697 @@
+  some h₁ + some h₂ = some (nonsingular_add h₁ h₂ $ λ h, (hx h).elim) :=
+by rw [← add_def, add, dif_neg hx]
+
+lemma some_add_some_of_X_ne' (hx : x₁ ≠ x₂) :
+  some h₁ + some h₂ = -some (nonsingular_add' h₁ h₂ $ λ h, (hx h).elim) :=
+some_add_some_of_X_ne h₁ h₂ hx
+
+end point
+
+end addition
+
+section group
+
+/-! ### The axioms for nonsingular rational points on a Weierstrass curve -/
+
+variables {F : Type u} [field F] {W : weierstrass_curve F} {x₁ x₂ y₁ y₂ : F}
+  (h₁ : W.nonsingular x₁ y₁) (h₂ : W.nonsingular x₂ y₂)
+
+include h₁
+
+lemma XY_ideal_neg_mul : XY_ideal W x₁ (C $ W.neg_Y x₁ y₁) * XY_ideal W x₁ (C y₁) = X_ideal W x₁ :=
+begin
+  have Y_rw :
+    (Y - C (C y₁)) * (Y - C (C (W.neg_Y x₁ y₁))) - C (X - C x₁)
+      * (C (X ^ 2 + C (x₁ + W.a₂) * X + C (x₁ ^ 2 + W.a₂ * x₁ + W.a₄)) - C (C W.a₁) * Y)
+      = W.polynomial * 1 :=
+  by linear_combination congr_arg C (congr_arg C ((W.equation_iff _ _).mp h₁.left).symm)
+    with { normalization_tactic := `[rw [neg_Y, weierstrass_curve.polynomial], C_simp, ring1] },
+  simp_rw [XY_ideal, X_class, Y_class, span_pair_mul_span_pair, mul_comm, ← _root_.map_mul,
+           adjoin_root.mk_eq_mk.mpr ⟨1, Y_rw⟩, _root_.map_mul, span_insert,
+           ← span_singleton_mul_span_singleton, ← mul_sup, ← span_insert],
+  convert mul_top _ using 2,
+  simp_rw [← @set.image_singleton _ _ $ adjoin_root.mk _, ← set.image_insert_eq, ← map_span],
+  convert map_top (adjoin_root.mk W.polynomial) using 1,
+  apply congr_arg,
+  simp_rw [eq_top_iff_one, mem_span_insert', mem_span_singleton'],
+  cases ((W.nonsingular_iff' _ _).mp h₁).right with hx hy,
+  { let W_X := W.a₁ * y₁ - (3 * x₁ ^ 2 + 2 * W.a₂ * x₁ + W.a₄),
+    refine ⟨C (C W_X⁻¹ * -(X + C (2 * x₁ + W.a₂))), C (C $ W_X⁻¹ * W.a₁), 0, C (C $ W_X⁻¹ * -1), _⟩,
+    rw [← mul_right_inj' $ C_ne_zero.mpr $ C_ne_zero.mpr hx],
+    simp only [mul_add, ← mul_assoc, ← C_mul, mul_inv_cancel hx],
+    C_simp,
+    ring1 },
+  { let W_Y := 2 * y₁ + W.a₁ * x₁ + W.a₃,
+    refine ⟨0, C (C W_Y⁻¹), C (C $ W_Y⁻¹ * -1), 0, _⟩,
+    rw [neg_Y, ← mul_right_inj' $ C_ne_zero.mpr $ C_ne_zero.mpr hy],
+    simp only [mul_add, ← mul_assoc, ← C_mul, mul_inv_cancel hy],
+    C_simp,
+    ring1 }
+end
+
+private lemma XY_ideal'_mul_inv :
+  (XY_ideal W x₁ (C y₁) : fractional_ideal W.coordinate_ring⁰ W.function_field)
+    * (XY_ideal W x₁ (C $ W.neg_Y x₁ y₁) * (X_ideal W x₁)⁻¹) = 1 :=
+by rw [← mul_assoc, ← fractional_ideal.coe_ideal_mul, mul_comm $ XY_ideal W _ _,
+       XY_ideal_neg_mul h₁, X_ideal,
+       fractional_ideal.coe_ideal_span_singleton_mul_inv W.function_field $ X_class_ne_zero W x₁]
+
+include h₂
+
+lemma XY_ideal_mul_XY_ideal (hxy : x₁ = x₂ → y₁ ≠ W.neg_Y x₂ y₂) :
+  X_ideal W (W.add_X x₁ x₂ $ W.slope x₁ x₂ y₁ y₂) * (XY_ideal W x₁ (C y₁) * XY_ideal W x₂ (C y₂))
+    = Y_ideal W (line_polynomial x₁ y₁ $ W.slope x₁ x₂ y₁ y₂)
+      * XY_ideal W (W.add_X x₁ x₂ $ W.slope x₁ x₂ y₁ y₂)
+        (C $ W.add_Y x₁ x₂ y₁ $ W.slope x₁ x₂ y₁ y₂) :=
```

Comment:
This should not require `nonsingular`, `equation` should be enough. I realized this when reading your proof in the draft, and I pushed a commit to remove the nonsingular conditions, and it still compiled. Worth a remark in the paper.